mautic/mautic
# TPROD-233 Enable hybrid Twig templates support in Mautic (`MauticCoreBundle:Default:content.html.twig`)

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | 4.x
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | yes
| BC breaks?                             | no
| Automated tests included?              | no
| Related user documentation PR URL      | N/A
| Related developer documentation PR URL | WIP
| Issue(s) addressed                     | Relates to #10852 




#### Description:

_First part of the Epic ["The Big PHP to Twig Template Engine migration"](https://mautic.atlassian.net/browse/TPROD-232)_

In Symfony 4, the PHP templating engine [has been officially deprecated](https://symfony.com/doc/4.4/templating/PHP.html), and has been removed in Symfony 5. This PR lays the groundwork for Twig support by introducing `MauticCoreBundle:Default:content.html.twig` which internal bundles and plugin authors can extend in preparation to the Symfony 5 migration 🎉

_Note that this PR should target the 4.x branch, as we'll need to migrate all internal bundles to Twig **before** we can go ahead and remove support for PHP templates (which will take place in the 5.x branch)._ 

The following modules are now available through Twig:
- `MauticCoreBundle:Default:content.html.twig` and all of its dependencies/children (`slim.html.twig, base.html.twig, pageheader.html.twig, navbar.html.twig, script.html.twig, head.html.twig, output.html.twig`)
- GlobalSearch
- Notifications
- Flashes
- Menus
- All Symfony custom form styles translated from PHP to Twig (`mautic_form_layout.twig`)
- `UserBundle:Views:Profile:index.html.twig`, which is the "My Account" page in Mautic

https://user-images.githubusercontent.com/17739158/162592213-337ad966-46b6-4704-a23d-f77835cbbfe8.mp4



#### Steps to test this PR:
1. Open this PR in Gitpod or locally
2. In the top right corner of the screen, click on your user name > "Account"
3. Click around and see if everything looks and works as expected
4. Click "Save". You should see a message saying "Your account has been updated."
5. Navigate to other screens. Everything should still work as expected. While other screens are not rendered through Twig yet, some of the main components _are_ rendered by Twig if the user is first on the "My Account" page.
6. Go to phpMyAdmin (https://mautic.ddev.site:8037/ in DDEV). Go to `db > notifications`. Click "SQL" on top of the screen. Execute the following query to create a new notification (if you're using a database prefix, make sure to add that before the `notifications` table name!)

```SQL
INSERT INTO `notifications` (`id`, `user_id`, `type`, `header`, `message`, `date_added`, `icon_class`, `is_read`) VALUES
(1, 1, NULL, 'Lovely header', 'Hello hello test', '2022-04-08 19:34:51', NULL, 0);
```

7. Go back to Mautic. Refresh the page. In the top left corner of the screen, click the notification bell. You should now see the notification you just created (see below for some example screenshots of notifications).



### Notification bell

Notification with no `iconClass` set (defaults to `fa-star`):

<img width="407" alt="image" src="https://user-images.githubusercontent.com/17739158/162518986-9d6452ee-ac6c-4e9e-94cb-d1d519c9d1c0.png">

Notification with `iconClass` set to `fa-shower`:

<img width="337" alt="image" src="https://user-images.githubusercontent.com/17739158/162521905-a5fbed39-aba4-407d-9e43-4023f91b2cf8.png">

Notification with `iconClass` set to `img:https://www.mautic.org/sites/default/files/2020-03/Mautic_Logo_LB.png`

<img width="333" alt="image" src="https://user-images.githubusercontent.com/17739158/162521250-19108065-d36e-4852-b33b-de544d2ae2c4.png">

Notification with header:

<img width="339" alt="image" src="https://user-images.githubusercontent.com/17739158/162523218-7baffc87-09c7-4168-b757-317795432085.png">

No notifications:

<img width="375" alt="image" src="https://user-images.githubusercontent.com/17739158/162525105-ee461e9c-eb57-4482-8fb8-6ea3e4555e4f.png">

Update available:

<img width="402" alt="image" src="https://user-images.githubusercontent.com/17739158/162525207-b225941c-19e0-4e0a-bc28-87e843c1e7cb.png">

### Search icon

<img width="390" alt="image" src="https://user-images.githubusercontent.com/17739158/162526594-f74220fe-c893-4fbb-b977-c3deed360a78.png">

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10129"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
